### ISSUE TITLE
Children process is leaked

### ISSUE THREAD

**Issue by ahmetalpbalkan** [Azure/WALinuxAgent#272]
(CentOS 6.5 with the develop branch of the agent but I'm not sure if the problem is limited to CentOS)

It appears like when the waagent service is stopped (`service waagent stop`) we're leaking the extension handler child process. Likely a signal forwarding issue.

> Sitenote: it also appears like I can run multiple `-daemon`s manually as well, what happened to our pidfile checking?

Here's pstree output from commands:

### service waagent start

```
  ├─python,30086 /usr/sbin/waagent -daemon
  │   ├─python,30108 /usr/sbin/waagent run-exthandlers
  │   ├─{python},30104
  │   └─{python},30106
```

### service waagent stop

> child process gets parented to init

```
$ pstree -a -p
init,1
  ...
  ├─python,30108 /usr/sbin/waagent run-exthandlers
  ...
```

### service waagent start

two instances of `run-exthandlers`

```
init,1
  ...
  ├─python,30108 /usr/sbin/waagent run-exthandlers
  ├─python,30159 /usr/sbin/waagent -daemon
  │   ├─python,30182 /usr/sbin/waagent run-exthandlers
  │   ├─{python},30178
  │   └─{python},30180
```

### service waagent stop

Another run-exthandlers parented to init.

```
init,1
  ...
  ├─portreserve,1297
  ├─python,30108 /usr/sbin/waagent run-exthandlers
  ├─python,30182 /usr/sbin/waagent run-exthandlers
  ├─rpc.statd,1365
```

### service waagent start

```
init,1
  ...
  ├─python,30108 /usr/sbin/waagent run-exthandlers
  ├─python,30182 /usr/sbin/waagent run-exthandlers
  ├─python,30213 /usr/sbin/waagent -daemon
  │   ├─python,30234 /usr/sbin/waagent run-exthandlers
  │   ├─{python},30230
  │   └─{python},30232
```

### service waagent stop

```
init,1
  ...
  ├─portreserve,1297
  ├─python,30108 /usr/sbin/waagent run-exthandlers
  ├─python,30182 /usr/sbin/waagent run-exthandlers
  ├─python,30234 /usr/sbin/waagent run-exthandlers
  ├─rpc.statd,1365
```

so this can keep going on...

**Comment by brendandixon** [comment]
@ahmetalpbalkan Hmm, DaemonHandler#main still checks for the pidfile. The child process does not overwrite that file (since it does not call DaemonHandler#main).

I was under the impression that killing the parent process would terminate the child??

**Comment by ahmetalpbalkan** [comment]
> Hmm, DaemonHandler#main still checks for the pidfile.

I can run just two `/usr/sbin/waagent -daemon`s in separate shells just fine on the same machine, that's why I thought the pidfile thing is not working.

> I was under the impression that killing the parent process would terminate the child??

me too, hence the issue. Python subprocess.Popen probably has some weird thing going on. I tried sending SIGTERM and SIGSEGV to parent process, in both cases the child got re-parented to init and just continued to run...

**Comment by brendandixon** [comment]
@ahmetalpbalkan Ok. I'll look into both of these once I complete my current tests.

**Comment by brendandixon** [comment]
@ahmetalpbalkan Should be fixed. I forgot to mention the issue in my PR.

**Comment by ahmetalpbalkan** [comment]
@brendandixon just tried #278. it does not fix the issue.

**Comment by ahmetalpbalkan** [comment]
fixed by #272.
